lambda_metadata.tf

```diff
@@ -0,0 +1,181 @@
+locals {
+  enable_metadata_purge_lambda = {
+    development = true
+    qa          = false
+    integration = false
+    preprod     = false
+    production  = false
+  }
+}
+
+
+resource "aws_iam_role" "metadata_removal_lambda" {
+  count = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  name  = "metadata-removal-lambda-role"
+  path  = "/"
+  assume_role_policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Action = "sts:AssumeRole"
+        Effect = "Allow"
+        Sid    = ""
+        Principal = {
+          Service = "lambda.amazonaws.com"
+        }
+      },
+    ]
+  })
+
+  tags = { Name = "metadata-removal-lambda-role" }
+}
+
+
+resource "aws_iam_policy_attachment" "metadata_removal_lambda" {
+  count      = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  name       = "metadata-removal-lambda"
+  policy_arn = aws_iam_policy.metadata_removal_lambda[count.index].arn
+  roles      = [aws_iam_role.metadata_removal_lambda[count.index].name]
+}
+
+data "aws_iam_policy_document" "metadata_removal_lambda" {
+  count = local.enable_metadata_purge_lambda[local.environment] ? 1 : 0
+  statement {
+    sid = "LogGroup"
+    actions = [
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents"
+    ]
+    resources = [
+      "*"
+    ]
+  }
+
+  statement {
+    sid = "EMRDescribeClusters"
+    actions = [
+      "elasticmapreduce:DescribeCluster"
+    ]
+    resources = ["*"]
+  }
+
+  statement {
+    sid = "HBaseRootDirOperations"
```

Comment:
I think this should be split out and the DeleteObject should only be on the specific prefix (i.e. `/data/hbase/meta_`). Just adds an extra layer that it can't delete anything it shouldn't - including the actual meta folder of `/data/hbase/meta`.

Reply:
You could build that as a local and pass it to lambda as well actually.

Reply:
I considered that, but wondered about the impact of root-dir changing (especially in lower envs, if less likely in higher)
- A replica started over one directory could end up trying to delete metadata from another directory and wouldn't find it
- Requires this terraform to be re-deployed on any root-dir changes in internal-compute and/or ingestion
If we don't need the lambda to work when launching replicas over previous hbase directories, then I suppose it doesn't matter so much

Reply:
But the root dir is still a remote state var here anyway, so changes in the root dir would already cause the above to happen?

This is more about not have delete as an option on `root_dir/*` but having it on a more specific dir. That dir is hardcoded in to the lambda anyway, so if that path itself changed then it would already stop working. I just don't think any component except HBase itself can have delete rights on the root dir, it feels too much of a risk.